File: app-python/entitlements_app/services/register_feature_service.py
"""Register a feature key that can be attached to plans and checked at runtime."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.register_feature_request import RegisterFeatureRequest
from entitlements_app.entities import Feature
from entitlements_app.ports import EntitlementsRepository


class RegisterFeatureService:
    """Service handling Command `RegisterFeature`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: RegisterFeatureRequest) -> Feature:
        # spec-driven body for Command RegisterFeature.
        # intent: Register a feature key that can be attached to plans and checked at runtime.
        # requires: EntitlementsRepository
        # return: Feature (entity); construct from request inputs and persist via EntitlementsRepository.saveFeature.
        value = Feature(tenantId=request.tenantId, featureKey=request.featureKey, displayName=request.displayName, valueType=request.valueType, defaultUnit=request.defaultUnit, active=request.active)
        return self.entitlements_repository.saveFeature(value)

File: app/src/main/java/com/example/spring/RegisterFeatureService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Register a feature key that can be attached to plans and checked at runtime.
 */
@Service
public class RegisterFeatureService {

    private final EntitlementsRepository entitlementsRepository;

    public RegisterFeatureService(EntitlementsRepository entitlementsRepository) {
        this.entitlementsRepository = entitlementsRepository;
    }

    @Transactional
    public Feature handle(RegisterFeatureRequest request) {
        // spec-driven body for RegisterFeature.
        // intent: Register a feature key that can be attached to plans and checked at runtime.
        // requires: EntitlementsRepository
        // return: Feature (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveFeature.
        Feature value = new Feature(request.tenantId(), request.featureKey(), request.displayName(), request.valueType(), request.defaultUnit(), request.active());
        return this.entitlementsRepository.saveFeature(value);
    }
}